File: budgetMicroservice/src/main/java/com/budgetMicroservice/dto/CustomPageableDTO.java
package com.budgetMicroservice.dto;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
@AllArgsConstructor
public class CustomPageableDTO {
    private int pageNumber;
    private int pageSize;
    private int offset;
    private boolean paged;
    private boolean unpaged;
    private SortDTO sort;

    public String getSortField() {
        return sort != null ? sort.getField() : null;
    }

    public String getSortOrder() {
        return sort != null && sort.getDirection() != null ? sort.getDirection().name() : "ASC";
    }
}

File: client/src/app/services/models/custom-pageable-dto.ts
import { SortDto } from '../models/sort-dto';

/**
 * DTO representing pagination details including page number, page size, offset, sort order, and correlation ID.
 */
export interface CustomPageableDto {

  /**
   * Unique correlation ID to trace and associate requests/responses.
   */
  correlationId: string;

  /**
   * Offset of the requested page in relation to the total data set.
   */
  offset: number;

  /**
   * Page number of the requested data.
   */
  pageNumber: number;

  /**
   * Page size specifying the number of items per page.
   */
  pageSize: number;

  /**
   * Indicates whether pagination is enabled.
   */
  paged: boolean;
  sort?: SortDto | null;

  /**
   * Indicates whether pagination is disabled (for unpaged data requests).
   */
  unpaged: boolean;
}
